### DISCUSSION TITLE
measure description

### DISCUSSION THREAD

**Discussion by kaem93** [microsoft/semantic-link-labs#383] [General]
Hi,

I've firstly run the module "connect_semantic_model" in order to connect to a semantic model and being able to use TOM package.
After that, I've run the "labs.tom.TOMWrapper.generate_measure_descriptions" command but obtained this error:

TypeError: TOMWrapper.generate_measure_descriptions() missing 1 required positional argument: 'self'

What is the missing argument? Based on documentation, no argument is mandatory (measure_name and max_batch_size are optional)

**Comment by m-kovalsky** [comment]
I'm not sure you are calling the function properly (even though I haven't seen your code). Here is the proper way to do this. Just enter in your semantic model name and workspace name in the corresponding parameters. You can also specify a measure or list of measures using the 'measure_name' parameter within the generate_measure_descriptions function.

```python
from sempy_labs.tom import connect_semantic_model

with connect_semantic_model(dataset='', workspace='', readonly=False) as tom:
    tom.generate_measure_descriptions()
```

**Reply by kaem93** [reply]
Yeah thank you I closed the discussion because I found an error in my script.
Happy holidays!

**Reply by m-kovalsky** [reply]
Great! Happy holidays!

**Reply by kaem93** [reply]
Just a last question: I run the script for all measures in my Dataset and, after few minutes, got this error:

FabricHTTPException: 401 Unauthorized for url: https://api.fabric.microsoft.com//explore/v202304/nl2nl/completions
Headers: {'Content-Length': '0', 'X-PowerBI-Error-Info': 'UserDoesNotBelongToCurrentCluster', 'X-PowerBI-Error-Details': "Current user belongs to cluster '', but this request was sent to cluster: 'WABI-WEST-EUROPE-D-PRIMARY-redirect.analysis.windows.net'.", 'request-redirected': 'true', 'home-cluster-uri': 'https://wabi-west-europe-redirect.analysis.windows.net/', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'Access-Control-Expose-Headers': 'RequestId,X-PowerBI-Error-Info,X-PowerBI-Error-Details', 'RequestId': 'e2f6e9df-42ee-4e0e-907d-86922c21cf86', 'Date': 'Tue, 24 Dec 2024 12:45:43 GMT'}

I tried the script for a specific measure (added in tom.generate_measure_descriptions(measure_name='mymeasure')) and it worked well; I re-run the same script but got the above error. 
Any idea?

**Comment by m-kovalsky** [comment]
You need to run this on a paid F64 sku or higher.

**Reply by kaem93** [reply]
The workspace is assigned to an F512 capacity

**Comment by m-kovalsky** [comment]
Hmm, interesting. I'll have to check and get back to you - likely after the holidays.

**Comment by m-kovalsky** [comment]
This should be resolved in [0.9.0](https://github.com/microsoft/semantic-link-labs/releases/tag/0.9.0).

**Comment by sabari** [comment]
hi m-Kovalsky
When i am trying to execute we are getting the below error .

FabricHTTPException: 401 Unauthorized for url: https://msitapi.fabric.microsoft.com//explore/v202304/nl2nl/completions
Headers: {'Cache-Control': 'no-store, must-revalidate, no-cache', 'Pragma': 'no-cache', 'Transfer-Encoding': 'chunked', 'Content-Type': 'application/octet-stream', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'RequestId': 'f6feff9b-b919-4916-a116-f284f5093edc', 'Access-Control-Expose-Headers': 'RequestId', 'Date': 'Fri, 04 Apr 2025 02:53:55 GMT'}

We are executing this in F256 capacity

**Reply by sabari** [reply]
@m-kovalsky - any update on this

**Comment by m-kovalsky** [comment]
Which region are you using?

**Comment by sabari** [comment]
West us

**Comment by m-kovalsky** [comment]
Do you get this error immediately or after the function runs for some time?

**Reply by sabari** [reply]
@m-kovalsky - We are getting this error after 2mins

FabricHTTPException: 500 Internal Server Error for url: https://msitapi.fabric.microsoft.com//explore/v202304/nl2nl/completions
Error: {"code":"JsonReaderException","message":"Unterminated string. Expected delimiter: \". Path '[4].description', line 25, position 126."}
Headers: {'Cache-Control': 'no-store, must-revalidate, no-cache', 'Pragma': 'no-cache', 'Transfer-Encoding': 'chunked', 'Content-Type': 'application/json; charset=utf-8', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains', 'X-Frame-Options': 'deny, deny', 'X-Content-Type-Options': 'nosniff, nosniff', 'RequestId': '4859730a-663b-49df-8ea5-8e6f2cb398a7', 'Access-Control-Expose-Headers': 'RequestId', 'Date': 'Fri, 18 Apr 2025 05:09:09 GMT'}

**Reply by sabari** [reply]
this  the code we are using it 

from sempy_labs.tom import connect_semantic_model

dataset = "" # Enter the name or ID of your semantic model
workspace = "" # Enter the name or ID of the workspace in which the semantic model resides
readonly = False # Set this to False to overwrite existing descriptions

with connect_semantic_model(dataset=dataset, workspace=workspace, readonly=readonly) as tom:
    x = tom.generate_measure_descriptions()

**Reply by m-kovalsky** [reply]
Ah so that’s a different error. I think I know the issue for that. Do you have measures which have DAX that contains double quotes?

**Reply by sabari** [reply]
`yes
`

**Reply by sabari** [reply]
@m-kovalsky  - any solution for this?

**Comment by m-kovalsky** [comment]
So I tested for measures with double quotes and they work fine. Are you able to identify the measure which causes the issue? You can specify a measure or list of measures within the function’s parameter. Sharing the DAX of the problem measure would help a lot.

**Reply by sabari** [reply]
it has more the 100+ measures. Is there is a way to find which DAX is not supported while using your code

Error Log:
2025-04-21 19:10:55,260 WARN ApacheHttpClient$DefaultServiceUnavailableRetryStrategy [autoTune-credential-refreshing]: ServiceUnavailableHandler: Deduced service unavailable. client-request-id=ccb48081-3ba7-453f-b104-b3d4ca5157ca Attempt=3 HttpStatusCode=500
2025-04-21 19:10:55,261 WARN ApacheHttpClient$DefaultServiceUnavailableRetryStrategy [autoTune-credential-refreshing]: ServiceUnavailableHandler will retry request. client-request-id=ccb48081-3ba7-453f-b104-b3d4ca5157ca Reason=ServiceUnavailable RetryDelay=25000ms
2025-04-21 19:11:20,485 WARN ApacheHttpClient$DefaultServiceUnavailableRetryStrategy [autoTune-credential-refreshing]: ServiceUnavailableHandler: Deduced service unavailable. client-request-id=ccb48081-3ba7-453f-b104-b3d4ca5157ca Attempt=4 HttpStatusCode=500
2025-04-21 19:11:20,485 WARN ApacheHttpClient$DefaultServiceUnavailableRetryStrategy [autoTune-credential-refreshing]: ServiceUnavailableHandler has no more retries. client-request-id=ccb48081-3ba7-453f-b104-b3d4ca5157ca TotalRetryAttempts=3
2025-04-21 19:11:20,487 ERROR TokenLibrary [autoTune-credential-refreshing]: An error occurred in proxyToSparkCoreWorkloadV2
java.io.IOException: {"code":500,"responseHeaders":{"Server":"Kestrel","x-ms-root-activity-id":"a97b03b3-48be-4839-a152-7bbc6c52a478","x-ms-routing-hint":"host001_sc-001","x-ms-gateway-request-id":"a8e96d07-acab-43c1-a54d-4c3cf14380ab | client-request-id : ccb48081-3ba7-453f-b104-b3d4ca5157ca","x-ms-workspace-name":"dfe9be0f-977c-4337-bb45-c2ab72a1ba6f","Date":"Mon, 21 Apr 2025 19:11:19 GMT","x-ms-activity-id":"a8e96d07-acab-43c1-a54d-4c3cf14380ab","Strict-Transport-Security":"max-age=31536000; includeSubDomains","Content-Length":"98","x-ms-workload-resource-moniker":"c178de27-ff7e-4b1e-aae1-04ea3e816265","x-ms-client-request-id":"ccb48081-3ba7-453f-b104-b3d4ca5157ca","x-ms-current-utc-date":"4/21/2025 7:11:20 PM","Content-Type":"application/json; charset=utf-8"},"message":"{\r\n  \"code\": \"InternalServerError\",\r\n  \"message\": \"Unexpected internal error, please try again\"\r\n}"}
	at com.microsoft.azure.trident.tokenlibrary.TokenLibrary.proxyToSparkCoreWorkloadV2(TokenLibrary.scala:1040)
	at com.microsoft.azure.trident.tokenlibrary.TokenLibrary.proxyToSparkCoreWorkload(TokenLibrary.scala:935)
	at com.microsoft.azure.trident.tokenlibrary.TokenLibrary$.proxyToSparkCoreWorkload(TokenLibrary.scala:1374)
	at com.microsoft.autotune.client.AutoTuneCredentialManager.getAutoTuneCredentialFromSparkCoreWorkload(AutoTuneCredentialManager.scala:92)
	at com.microsoft.autotune.client.AutoTuneCredentialManager.$anonfun$fetchAutoTuneSasUrl$1(AutoTuneCredentialManager.scala:78)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at com.microsoft.autotune.common.util.AutotuneUtils$.timeTracking(AutotuneUtils.scala:60)
	at com.microsoft.autotune.client.AutoTuneCredentialManager.fetchAutoTuneSasUrl(AutoTuneCredentialManager.scala:77)
	at com.microsoft.autotune.client.AutoTuneCredentialManager.refreshCredential(AutoTuneCredentialManager.scala:50)
	at com.microsoft.autotune.client.AutoTuneCredentialManager.$anonfun$new$1(AutoTuneCredentialManager.scala:33)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
2025-04-21 19:11:20,731 WARN ApacheHttpClient$DefaultServiceUnavailableRetryStrategy [autoTune-credential-refreshing]: ServiceUnavailableHandler: Deduced service unavailable. client-request-id=ba45358e-6c66-4dbd-87df-5089f28bc2c4 Attempt=1 HttpStatusCode=500
2025-04-21 19:11:20,732 WARN ApacheHttpClient$DefaultServiceUnavailableRetryStrategy [autoTune-credential-refreshing]: ServiceUnavailableHandler will retry request. client-request-id=ba45358e-6c66-4dbd-87df-5089f28bc2c4 Reason=ServiceUnavailable RetryDelay=10000ms
2025-04-21 19:11:30,989 WARN ApacheHttpClient$DefaultServiceUnavailableRetryStrategy [autoTune-credential-refreshing]: ServiceUnavailableHandler: Deduced service unavailable. client-request-id=ba45358e-6c66-4dbd-87df-5089f28bc2c4 Attempt=2 HttpStatusCode=500
2025-04-21 19:11:30,989 WARN ApacheHttpClient$DefaultServiceUnavailableRetryStrategy [autoTune-credential-refreshing]: ServiceUnavailableHandler will retry request. client-request-id=ba45358e-6c66-4dbd-87df-5089f28bc2c4 Reason=ServiceUnavailable RetryDelay=17000ms

**Comment by m-kovalsky** [comment]
Could you try running this? It gets descriptions for 5 measures at a time. Before it runs the API to get the descriptions, it prints out the 5 measure names. Find the last printout of the measure names and within those you will find the culprit. Would then be great to see the DAX for those 5 measures (you can then run those 5 measures individually to find the specific culprit).

```python
from sempy_labs.tom import connect_semantic_model
import pandas as pd
from typing import Optional, List
import sempy_labs._icons as icons

dataset = ''
workspace = ''

with connect_semantic_model(dataset=dataset, workspace=workspace, readonly=True) as tom:

    def generate_measure_descriptions(
            measure_name: Optional[str | List[str]] = None,
            max_batch_size: Optional[int] = 5,
        ) -> pd.DataFrame:
            """
            Auto-generates descriptions for measures using an LLM. This function requires a paid F-sku (Fabric) of F64 or higher.
            Setting the 'readonly' parameter in connect_semantic_model to True will allow you to see the auto-generated descriptions in a dataframe. Setting the 'readonly' parameter
            to False will update the descriptions for the measures within the 'measure_name' parameter.

            Parameters
            ----------
            measure_name : str | List[str], default=None
                The measure name (or a list of measure names).
                Defaults to None which generates descriptions for all measures in the semantic model.
            max_batch_size : int, default=5
                Sets the max batch size for each API call.

            Returns
            -------
            pandas.DataFrame
                A pandas dataframe showing the updated measure(s) and their new description(s).
            """

            df = pd.DataFrame(
                columns=["Table Name", "Measure Name", "Expression", "Description"]
            )
            data = []

            # import concurrent.futures
            if measure_name is None:
                measure_name = [m.Name for m in tom.all_measures()]

            if isinstance(measure_name, str):
                measure_name = [measure_name]

            if len(measure_name) > max_batch_size:
                measure_lists = [
                    measure_name[i : i + max_batch_size]
                    for i in range(0, len(measure_name), max_batch_size)
                ]
            else:
                measure_lists = [measure_name]

            # Each API call can have a max of 5 measures
            for measure_list in measure_lists:
                payload = {
                    "scenarioDefinition": {
                        "generateModelItemDescriptions": {
                            "modelItems": [],
                        },
                    },
                    "workspaceId": tom._workspace_id,
                    "artifactInfo": {"artifactType": "SemanticModel"},
                }
                for m_name in measure_list:
                    expr, t_name = next(
                        (ms.Expression, ms.Parent.Name)
                        for ms in tom.all_measures()
                        if ms.Name == m_name
                    )
                    if t_name is None:
                        raise ValueError(
                            f"{icons.red_dot} The '{m_name}' measure does not exist in the '{tom._dataset_name}' semantic model within the '{tom._workspace_name}' workspace."
                        )

                    new_item = {
                        "urn": m_name,
                        "type": 1,
                        "name": m_name,
                        "expression": expr,
                    }
                    payload["scenarioDefinition"]["generateModelItemDescriptions"][
                        "modelItems"
                    ].append(new_item)

                print(measure_list)

                response = _base_api(
                    request="/explore/v202304/nl2nl/completions",
                    method="post",
                    payload=payload,
                )

                for item in response.json().get("modelItems", []):
                    ms_name = item["urn"]
                    if ms_name.startswith("urn: "):
                        ms_name = ms_name[5:]
                    desc = item.get("description")
                    (table_name, expr) = next(
                        (m.Parent.Name, m.Expression)
                        for m in tom.all_measures()
                        if m.Name == ms_name
                    )
                    tom.model.Tables[table_name].Measures[ms_name].Description = desc

                    # Collect new descriptions in a dataframe
                    new_data = {
                        "Table Name": table_name,
                        "Measure Name": ms_name,
                        "Expression": expr,
                        "Description": desc,
                    }

                    data.append(new_data)

            if data:
                df = pd.concat([df, pd.DataFrame(data)], ignore_index=True)

            return df

    x = generate_measure_descriptions()
```

**Reply by sabari** [reply]
@m-kovalsky with this logic i checked where ever we have "" this code is not working. In real world scenario this will not workout. If you can provide some alternate solutions will be good to use.

**Reply by m-kovalsky** [reply]
That’s odd as I tried this and it works for me. Would you be able to send one of the DAX statements which give an error?

**Reply by sabari** [reply]
@m-kovalsky 

VAR tVar =
    IF (
        CONTAINS ( 'dimproduct', 'dimproduct'[ProductGroup], "Accessories" ),
        [Attach_Int_Fcst]
    )
RETURN
    tVar
    
    i am also working in Microsoft as vendor(v-sabgup)
 if possible we can connect to rectify this. That would be more helpful to implement across the portfolio

**Reply by m-kovalsky** [reply]
I cannot reproduce this error. I made a measure with essentially the same code and it works just fine. In which environment are you running this (prod/msit etc.)? And in which region?

![image](https://github.com/user-attachments/assets/29f84800-f38a-47f0-b52a-1ec3476e6cf8)

**Reply by sabari** [reply]
@m-kovalsky using msit and west us

**Comment by sabari** [comment]
In msit

**Comment by sabari** [comment]
In msit and west us

**Comment by michael-head-leeco** [comment]
I am also getting a "401 Unauthorized for url: https://api.fabric.microsoft.com//explore/v202304/nl2nl/completions" error when attempting to generate metric descriptions.  We have an F64,  we have all necessary AI and Open AI configurations configured.  I have tried to just generate 5 metric descriptions at a time, but get the error.  I am able to connect to the semantic model and see the metrics.  The user executing the code is the owner of the semantic model and apart of the security group that has Fabric AI permissions.  The only issue I'm curious about is that our Capacity is in a different environment than our Power BI; F64 is in South and Power BI is in North Central. Could that cause any issues?  Here is the error message: "Specific measures approach failed: 401 Unauthorized for url: https://api.fabric.microsoft.com//explore/v202304/nl2nl/completions
Headers: {'Content-Length': '0', 'X-PowerBI-Error-Info': 'UserDoesNotBelongToCurrentCluster', 'X-PowerBI-Error-Details': "Current user belongs to cluster 'https://wabi-us-north-central-g-primary-redirect.analysis.windows.net/', but this request was sent to cluster: 'WABI-US-NORTH-CENTRAL-redirect.analysis.windows.net'.", 'request-redirected': 'true', 'home-cluster-uri': 'https://wabi-us-north-central-g-primary-redirect.analysis.windows.net/', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'Access-Control-Expose-Headers': 'RequestId,X-PowerBI-Error-Info,X-PowerBI-Error-Details', 'RequestId': '91e17eb4-ed71-499e-b884-70e4d3ebfd89', 'Date': 'Wed, 09 Jul 2025 21:21:28 GMT'}"

**Comment by ecampuss** [comment]
Hi @m-kovalsky,

I am trying to run the tom.generate_measure_descriptions function, but I am getting the following error message below:

FabricHTTPException: 401 Unauthorized for url: https://api.fabric.microsoft.com//explore/v202304/nl2nl/completions
Headers: {'Content-Length': '0', 'X-PowerBI-Error-Info': 'UserDoesNotBelongToCurrentCluster', 'X-PowerBI-Error-Details': "Current user belongs to cluster 'https://wabi-west-us-e-primary-redirect.analysis.windows.net/', but this request was sent to cluster: 'WABI-WEST-US-B-PRIMARY-redirect.analysis.windows.net'.", 'request-redirected': 'true', 'home-cluster-uri': 'https://wabi-west-us-e-primary-redirect.analysis.windows.net/', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'Access-Control-Expose-Headers': 'RequestId,X-PowerBI-Error-Info,X-PowerBI-Error-Details', 'RequestId': 'd3505b03-5dac-4df4-8a21-d510366d6974', 'Date': 'Mon, 18 Aug 2025 20:09:18 GMT'}

Can you help me with that? I am running the function using the Fabric Notebook under PBI Service. Capacity that we are using are F64.

**Reply by CharlesSaulnier** [reply]
Hi, I run into the same issue; we're not yet on F64, we have a P1 that will be migrated to F64 in the coming months.

Detailed error:

FabricHTTPException: 401 Unauthorized for url: https://api.fabric.microsoft.com//explore/v202304/nl2nl/completions
Headers: {'Content-Length': '0', 'X-PowerBI-Error-Info': 'UserDoesNotBelongToCurrentCluster', 'X-PowerBI-Error-Details': "Current user belongs to cluster 'https://wabi-canada-central-b-primary-redirect.analysis.windows.net/', but this request was sent to cluster: 'WABI-CANADA-CENTRAL-redirect.analysis.windows.net'.", 'request-redirected': 'true', 'home-cluster-uri': 'https://wabi-canada-central-b-primary-redirect.analysis.windows.net/', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'Access-Control-Expose-Headers': 'RequestId,X-PowerBI-Error-Info,X-PowerBI-Error-Details', 'RequestId': '4530449c-6c8f-44f9-be45-57da282b098a', 'Date': 'Wed, 20 Aug 2025 12:15:06 GMT'}

Both our Azure tenant and P1 capacity are in canada central region

**Reply by CharlesSaulnier** [reply]
Hi @m-kovalsky, any update on that front? I tested the function again this morning on a new capacity (in the same region), and still get the 401 error. If there was a parameter to manage the cluster the request is sent to, or any way to validate the result of the redirect, that would be awesome

**Reply by ecampuss** [reply]
I am also looking for an update too and I have the same questions as you Charles. Hope we have a way to define the cluster or, at least, understand the error message, which does not make much sense for me.

**Reply by ecampuss** [reply]
@m-kovalsky

**Reply by m-kovalsky** [reply]
Do you have copilot enabled on your fabric capacity?

**Comment by ecampuss** [comment]
Yes, I have and the auto description option works fine if I open the model
through the Semantic Model and click on the create copilot's description
button.


On Mon, Sep 8, 2025 at 1:52 PM m-kovalsky ***@***.***> wrote:

> Do you have copilot enabled on your fabric capacity?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/semantic-link-labs/discussions/383#discussioncomment-14342688>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWLEGOMU4WICJXHRIF3F4MT3RWX37AVCNFSM6AAAAABUETRAXCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTIMZUGI3DQOA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: <microsoft/semantic-link-labs/repo-discussions/383/comments/14342688@
> github.com>
>


-- 
*Erick Campos*
Phone: (12) 98818-2337
MS Teams: erick.de.campos

**Reply by CharlesSaulnier** [reply]
Same here, Copilot is enabled, I can even use it with %chat-magics to suggest measure descriptions in the same notebook I use with tom.generate_measure_descriptions; ideally I'd like to leverage the fact tom.generate_measure_descriptions can push these descriptions to the model directly

**Reply by m-kovalsky** [reply]
Would you try [0.12.2](https://github.com/microsoft/semantic-link-labs/releases/tag/0.12.2) and let me know.

**Reply by CharlesSaulnier** [reply]
I no longer get the 401 error, thank you very much for that @m-kovalsky !

**Comment by ecampuss** [comment]
After upgrading the version of the library to the 0.12.2, it has executed
successfully in Fabric! Thank you so much for fixing that issue @m-kovalsky

On Fri, Sep 12, 2025 at 7:16 AM m-kovalsky ***@***.***> wrote:

> Would you try 0.12.2
> <https://github.com/microsoft/semantic-link-labs/releases/tag/0.12.2> and
> let me know.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/semantic-link-labs/discussions/383#discussioncomment-14383081>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWLEGOIWLP4CKOGUK4Z7HG33SKMQ3AVCNFSM6AAAAABUETRAXCVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTIMZYGMYDQMI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: <microsoft/semantic-link-labs/repo-discussions/383/comments/14383081@
> github.com>
>


-- 
*Erick Campos*
Phone: (12) 98818-2337
MS Teams: erick.de.campos

**Comment by sabari** [comment]
HI @m-kovalsky,

Now i am getting different error could you please check it
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[10], line 8
      5 readonly = False # Set this to False to overwrite existing descriptions
      7 with connect_semantic_model(dataset=dataset, workspace=workspace, readonly=readonly) as tom:
----> 8     x = tom.generate_measure_descriptions()

File /nfs4/pyenv-8845f6c3-2d39-41a6-b4bb-470832d2ba40/lib/python3.10/site-packages/sempy_labs/tom/_model.py:4673, in TOMWrapper.generate_measure_descriptions(self, measure_name, max_batch_size)
   4667 response = requests.post(
   4668     f"{prefix}/explore/v202304/nl2nl/completions",
   4669     headers=headers,
   4670     json=payload,
   4671 )
   4672 if response.status_code != 200:
-> 4673     raise FabricHTTPException(
   4674         f"Failed to retrieve descriptions: {response.text}"
   4675     )
   4677 for item in response.json().get("modelItems", []):
   4678     ms_name = item["urn"]

File /nfs4/pyenv-8845f6c3-2d39-41a6-b4bb-470832d2ba40/lib/python3.10/site-packages/sempy/fabric/exceptions/_exceptions.py:24, in FabricHTTPException.__init__(self, response)
     21 self.error_reason: str
     22 # The following section is copied from response.raise_for_status()
     23 # ----------------------------------------------------------------
---> 24 if isinstance(response.reason, bytes):
     25     # We attempt to decode utf-8 first because some servers
     26     # choose to localize their reason strings. If the string
     27     # isn't utf-8, we fall back to iso-8859-1 for all other
     28     # encodings. (See PR #3538)
     29     try:
     30         self.error_reason = response.reason.decode("utf-8")

AttributeError: 'str' object has no attribute 'reason'

**Reply by CharlesSaulnier** [reply]
I did get the same error when trying to call the function with the default parameters, but it should work with something like this:

`
from sempy_labs.tom import connect_semantic_model

with connect_semantic_model(dataset=dataset, readonly=True, workspace=workspace) as tom:
    for m in tom.all_measures():
    #    print(m.Name)
        measure = m.Name
        description = tom.generate_measure_descriptions(measure_name=measure)
        name = description["Measure Name"]
        measure_description = description["Description"]
        print(name + ': ' + measure_description)
`

**Comment by ecampuss** [comment]
What are you sending as dataset to the function?

Em sex., 12 de set. de 2025, 20:27, CharlesSaulnier <
***@***.***> escreveu:

**Reply by CharlesSaulnier** [reply]
A dataset's name from the workspace in context, as I said worst case I iterate over the measures and it works ok

**Comment by m-kovalsky** [comment]
Please try [0.12.3](https://github.com/microsoft/semantic-link-labs/releases/tag/0.12.3).
